### DISCUSSION TITLE
Integrating MIDI Tuning Standard (real-time retuning message) in GO

### DISCUSSION THREAD

**Discussion by 000masa000** [GrandOrgue/grandorgue#1395] [Developers]
Hi, I am interested to implement MIDI Tuning Standard real-time sysex messages into GrandOrgue. Fluidsynth / jOrgan has built-in support and it allows any tuning to be played in real-time, no matter how many notes. MIDI Notes become keys to a tuning map, which is dynamically updated with a sysex message just before the next noteon is played. Here is the info:
described in detail here:
https://www.midi.org/specifications/midi1-specifications/general-midi-specifications/general-midi-2/midi-tuning-updated

It might be nice to include support for the Oddsound MTS-ESP Master/Client retuning system which facilitates system-wide integration of tunings. They have a free master plugin that allows various scale formats to be loaded and applied at once to any connected software.
https://oddsound.com/devs.php
https://github.com/ODDSound/MTS-ESP

**Comment by 000masa000** [comment]
fluidsynth implementation is here at line 1966:
https://github.com/FluidSynth/fluidsynth/blob/1fc6f0935bacc6fb3e38b7bebe8b24f6cf110dab/src/synth/fluid_synth.c

**Comment by oleg68** [comment]
@larspalo What is your opinion about implementing midi tuning? Seems it might be useful for your recent sampleset.

**Comment by midimusic** [comment]
It looks like a good idea but would involve quite a lot of work - even to fully understand exactly what the requirement is. The links above are not particularly useful and a lot more investigation would be needed before a line of code was written. GO does not use fluidsynth.

How many users would use it even if it were available? Some sort of poll may be useful here.

csw900

**Reply by 000masa000** [reply]
Thanks for the reply! fluidsynth is obviously a different beast, but it is one of the open-source projects that has a really long-standing and stable implementation of the MTS (MIDI Tuning Standard) messages. The technique would likely be similar, as it provides (in the end) information about the nearest MIDI note and the amount of retuning upward not exceeding a semitone.

**Comment by larspalo** [comment]
@oleg68 As the request is written I'm very doubtful. All of that can be done with our custom temperaments. A way to in real time affect tuning is rather un orthodox regarding real pipe organs. To me it seems that the OP doesn't understand what GO is.

**Reply by 000masa000** [reply]
I understand that this feature would not be replicating standard organ keyboards. But it would (for example) allow new kinds of keyboards (like the Lumatone) to play tunings with more than 12 notes per octave, which GO temperaments do not currently allow. The "real-time retuning" name does not mean "bending" the pitch which I agree would be out of scope.

One nice historical example is Nicola Vicentino's Arciorgano, recently rebuilt in Basel by the early music department and performed by Johannes Keller (36 notes per octave across six ranks with variously split keys). Another is the organ built to Adrian Fokker's specifications in the Concertgebouw (31 tones per octave with a special keyboard).
The big advantage of a real time MTS tuning is that the problem of polyphony across MIDI channels is very elegantly handled. The incoming sysex message identifies itself as a real-time MTS message, then provides:
a note number (acting as a "key" to the data)
the nearest MIDI pitch not exceeding the desired frequency 
two bytes of retuning data (expressing a fraction of the distance to the next MIDI key

then a standard note on message with the "key" as its MIDI note number accesses the other three bytes of actual "sounding" information -- for example the incoming keys can be ANY note numbers between 0 and 127; so 127 notes of microtonal polyphony per MIDI channel can be handled

Notice the similarity to GO's current method of storing original tuning and equal temperament correction.

**Reply by larspalo** [reply]
Actually, it's quite possible already to play any number of keys an octave as the MIDI mapping of the keys is completely customizable in GO - but it would no longer adhere to the normal MIDI standard. Sub semi tones are a historic appendix that can be very interesting in many ways - but ultimately it was deemed not really useful for musicians and instrument builders.

For the custom mapping to be really useful you should also have a console MIDI keyboard that is actually designed to have its keys laid out like the keyboard model to get the real experience too.

For the Norrfjärden sample set @kerkovits lifted another very sensible question in #1379 - how can I at will choose to play d# instead of eb? I can solve it on the sounding level of ranks - but not with the MIDI signals (Note on/off for say key 51) activating the on screen keyboard(s). So, for a custom physical MIDI console that actually had d# and eb it would be possible, but I've not figured out how to do it with a normal MIDI keyboard.

**Reply by 000masa000** [reply]
Good to know, I will try this with my on-screen MIDI controller interface:
https://www.plainsound.org/hex
Using a mapping method I can play some of the special tunings I have there, but then I'll need to make the custom mappings for each new number of notes.... A good work-around for now!
The MTS method is so convenient and setup-free (with fluidsynth and other synths that support it), I would love to have it working with a sophisticated organ like GO, since a soundfont is just not detailed enough for real musical applications.

**Reply by 000masa000** [reply]
A colleague of mine is working on renderings of Bach organ repertoire in 1/5 comma extended Meantone (differentiating sharps and flats) and it would be ideal to be able to access the desired subsemitone on the fly as the music changes without constant patch changes.... MTS would allow this at the touch of a CC switch for each key that needs a retuning....

**Reply by 000masa000** [reply]
I have implemented 4 samples from the Norrfjarden church Hv8 Principal stop as a simple built-in looping sample on my app: plainsound.org/hex as "Baroque Organ Principal" in case anyone wants to get a feel of how it is to make an organ sound using this kind of alternate controller. I would recommend one of the extended Meantone tunings for trying it.

**Comment by 000masa000** [comment]
If no one on the current dev team is up for an MTS implementation, would there be anyone willing to help me get started on coding an implementation in my own fork of GrandeOrgue? Some advice on standard practices in this project would help me a lot. I am very motivated to be able to play many-note just intonation tunings with Lars Palo's recent Norrfjärden Church sample set, it sounds so beautiful.

**Reply by midimusic** [reply]
@000masa000 

I have never been able to make much sense of jorgan and so have never used it. I have just had another look at the latest version and it is even more opaque than it used to be. However I am interested in midi tuning as it is part of the midi standard - again I have never used it because I have never found any need for it.

However eplayOrgan DOES use fluidsynth and the organ is easy to use and understand. It does not currently implement midi tuning. However I suspect it would be relatively easy to implement the midi commands required. BUT there may be a problem as it currently cannot receive very long (more than about 20 bytes) real time sysex messages. Does midi tuning require the use of such long messages? Please attach an example of the longest sysex message that has to be accommodated.

I have just tried your "hex" link - it tries to send midi messages but does not say where they are going so I did not allow them.

I am skeptical about how normal manuals could be used to play other than 12 notes per octave. I also do not think the midi tuning standard will accommodate them. The whole of midi is based on 12 notes per octave and as I understand it midi tuning can  only change 
the tuning of each note by less than 1 semitone. I could easily be wrong here and would welcome any useful comments or information.

csw900

**Reply by 000masa000** [reply]
Thanks for the detailed reply! If you dare to try my app, the scary warning is simply because it generates sysex messages, and it ONLY does little MTS ones, and ONLY if you choose it in the sidebar, so saying OK to MIDI won't do anything scary :) That said, Edge and Chrome behave well, Firefox soso, Safari not.

Regarding the MTS message, it is basically short, as long as only one note is being sent at a time, which is how I have implemented it. The message, in decimal numbers is:

this.midi_output.send([240, 127, 0, 8, 2, 0, 1, this.mts[0], this.mts[1], this.mts[2], this.mts[3], 247]);

240 is sysex, 127 is real-time MTS, 0 is device ID (could be 127 also, which means "all devices", but not every program recognises this), 8 is the sub id for MTS, 2 is the MTS type (real-time single note retuning), 0 is the tuning map number, 1 is the number of notes being sent, and the mts array has four bytes, made up of note-to-assign-retuning, nearest-note-not-exceeding, fraction-of-semitone-MSB, fraction-of-semitone-LSB, 247 is sysex end

this is sent to the synth, and then the note-on message to note number "mts[0]" will produce the frequency computed by combining note number "mts[1]" and an upward pitch-bend applied from "mts[2]" and mts[3]" up to a maximum of (16383/16384)*100 cents. Therefore, the resolution exceeds both standard pitch bend (since this applies up AND down modification) or MPE (for the same reason).

Hope this makes sense? It is basically really easy. I am just not sure where in GrandeOrgue this should go.

**Reply by 000masa000** [reply]
It could also work fine with any 12 note temperament. The advantage is that you can even play many versions of closely tuned notes, or change temperaments on the fly, and still have 128 notes of potential polyphony....

**Reply by midimusic** [reply]
@000masa000 
Very useful information.
In order to fully implement the midi tuning standard into GO you will need to implement a GO version of all of the fluidsynth functions listed below:
int 	fluid_synth_activate_key_tuning (fluid_synth_t *synth, int bank, int prog, const char *name, const double *pitch, int apply)
 	Set the tuning of the entire MIDI note scale.
 int 	fluid_synth_activate_octave_tuning (fluid_synth_t *synth, int bank, int prog, const char *name, const double *pitch, int apply)
 	Activate an octave tuning on every octave in the MIDI note scale.
 int 	fluid_synth_activate_tuning (fluid_synth_t *synth, int chan, int bank, int prog, int apply)
 	Activate a tuning scale on a MIDI channel.
 int 	fluid_synth_deactivate_tuning (fluid_synth_t *synth, int chan, int apply)
 	Clear tuning scale on a MIDI channel (use default equal tempered scale).
 int 	fluid_synth_tune_notes (fluid_synth_t *synth, int bank, int prog, int len, const int *keys, const double *pitch, int apply)
 	Set tuning values for one or more MIDI notes for an existing tuning.
 int 	fluid_synth_tuning_dump (fluid_synth_t *synth, int bank, int prog, char *name, int len, double *pitch)
 	Get the entire note tuning for a given MIDI bank and program.
 int 	fluid_synth_tuning_iteration_next (fluid_synth_t *synth, int *bank, int *prog)
 	Advance to next tuning.
 void  fluid_synth_tuning_iteration_start (fluid_synth_t *synth)
 	Start tuning iteration.
The sysex tuning events (and I think also some additional control change events) will get into GO via RtMidi and will be processed in real time in the RtMidi Callback Buffer using the appropriate functions. You will need to distinguish between real time and non real time functions and process them accordingly.

Moving on to practical considerations - some of the incoming midi events (especially real time tuning events) may need to be recorded in any midi file which is produced. Exactly what needs to be recorded in a midi file? Does the organ's tuning have to be set up before playing the file (inconvenient) or will all of the necessary tuning events be recorded in the file so that it will always play with the correct tunings (possibly on another organ). Do you have any example midi files which use midi tuning? I have viewed thousands of midi files and never knowingly found one which contained any tuning events.

I am now going to do some more experiments with your hex app and hopefully view its output with midi-ox.

csw900

**Reply by midimusic** [reply]
@000masa000 
Using 64 bit Windows Edge - having previously told hex that I do not want midi output it has remembed it and subsequently does not ask the question again. There is no obvious way to reset it back to asking the question. Thus it will not allow setting up a midi output and consequently does not give any midi output. How does one reset it?

Using 32 bit Windows Edge on another computer the question is asked and "yes" I do want midi output - it allows me to set up a midi output port. However there is no midi output sent to this port when the type of scale is changed. The only midi output I can get is "All off" (on all 16 channels) when hex is closed by shutting down the browser.

I have also looked at Scala. It is so complex that I have not even been able to begin to get anything useful out of it. No organist would ever use this to tune an organ.

I have also seen a video on Youtube using "Midi Tuning Standard Tool" but this appears to be unavailable for download. Is it available and if so is it any good?

Thus all progress is negative. There is currently no easy way for an organist to tune an organ even if the organ was tuneable. Any ideas?

csw900

**Reply by 000masa000** [reply]
I also use Windows (64-bit) and Edge, I think it is just a matter of doing a hard reset (control-F5) and maybe emptying cache? I should probably figure out a way of having Edge ask each time, right now it stores the choice made previously by its own accord. Easiest way around it is to use an "in private" window by clicking on the three dots at the right side. Then it always asks!

It's not something I have in the code, it is just default behaviour of Chrome / Edge.

When scales are changed, I am not sending a tuning map. The way I do it is simply one note at a time, ONLY when the midi mapping is set to one of the "MTS" options. So the notes get tuned one at a time, which allows any number of notes to be played (doesn't rely on a preloaded map, instead it updates it dynamically). Scala is maddeningly complicated, but the file format is basic human readable.

Otherwise my plain MIDI output options are designed for softsynths that have built-in retuning (like Pianoteq).

You are right, Scala program is confusingly complicated. But the file format is simple and human-readable:
Scala file just looks like (! means comment):

!myScale.scl
A 12-tone scale, this is just a long description on one wrapping line, and the next line will say how many notes in my scale, which I am naming as "cents" from C; I start from the second note (C#), and go up to the octave, since C is assumed and is given by the Reference Frequency setting. It is also possible to reference from another note of course.
12
112.
204.
316.
386.
498.
582.
702.
814.
884.
996.
1088.
1200.


that's it, for any 12-note scale; you can also put ratios for just intervals (i.e. 5/4 instead of 386.)

The MTS-ESP tool is a new addition to the mix, it recognises sysex MTS messages and sends them to any soft synth that is willing to listen. Works well.  It is on GitHub and can be integrated rather easily into a project, apparently, but I didn't try yet. Here is the link:
https://github.com/ODDSound/MTS-ESP

**Reply by midimusic** [reply]
@000masa000 

I have got hex to open properly now - even private browsing did not work until I told edge to unblock the midi output! (using the icon at top of screen).
However I have still not been able to get any midi output from it even with MTS tuning selected. Do I need to connect a keyboard and play notes in order to get output? Originally It listed my keyboard input port but now no longer lists it even though a keyboard is connected to the system and turned on. Maybe another Edge problem.

I am not convinced that tuning one note at a time is a practical option which organists would use.

csw900

**Reply by 000masa000** [reply]
Thanks for sticking with me! I also use Windows. To get MIDI output you need to :
1) test for sound first: make sure "Sample Synth" in ON under "Sound Synthesis" and choose an instrument. Play notes by touching the screen if you have a touch screen If you don't, please close the sidebar by clicking the blue strip on the left side, then try using your computer keyboard. One of these two methods should generate sound through your computer output device. 
2) Now, change the Synthesis to "MIDI Output". Choose an Out Port to your sound generator. If the Sound generator is on your computer, I recommend using Tobias Erichsen's excellent Loop MIDI freeware. It allows you to create multiple loopback MIDI ports on your system that are available to any and all MIDI aware applications. If you want to test simple MIDI output, please select the MIDI mapping "all notes on selected channel, around MIDI note 60", to test normal noteons where C = C. Choose a 12 note tuning. This should enable the app to play notes normally.
1) select a MIDI IN port if you have a keyboard attached; because of Windows stealing MIDI drivers and making them exclusive, you need to be careful about other MIDI programs grabbing your keyboard driver first. I recommend o install Tobias Erichsen's LoopMIDI which is excellent and free, and make a few Loop MIDI ports on your system. Then use MIDI-OX FIRST (before starting browser or any other programs that might steal your keyboard MIDI driver) and use it to route your keyboard to LoopMIDI Port 1. To do this, View/MIDI Devices, choose to automatically attach Inputs to Output (checkbox), This makes your keyboard available system wide to all apps that want it, simply by selecting LoopMIDI port 1 as input. (It emulates MacOS 's CoreMIDI and IAC Port structure.)
![image](https://user-images.githubusercontent.com/34196907/221150866-2c0b12b0-4700-435b-827e-b0a3f7cb1ff0.png)
Now you can choose Loop MIDI Port 1 as an input wherever you like, including in my hex app. Make sure MIDI Output goes to a different port, or it will loop! If that happens, just refresh the page :) 

Now it should work, routing your input keyboard THROUGH the selected tuning to your output.

Finally, to get MTS messages, change the MIDI Mapping. Now the Note on messages will have NOTHING to do with the pitches you play, they will simply cycle in increasing order through all the available MIDI notes, which are retuned by the preceding (automatically generated) MTS Real-Time Sysex message. All of this should be invisible, to a user it sounds like magic retuning to whatever tuning with whatever number of notes, with visual on-screen feedback while playing-

**Reply by midimusic** [reply]
@000masa000 
Thanks for your help, at last I have got hex working and have viewed the MTS messages on midi-ox. There is still one possible bug (somewhere) - when I first got it going it gave its output directly I pressed an input note but later the input seemed to have a buffer that needed 9 or 10 notes entered before there was any output - then all 9 or 10 outputs occured at once.

However it is potentially a useful tuning tool which no-doubt does what you want at present. For further development I think you need to have a button labled "Tune" which outputs all 128 tuning sysex's in one go. (Keeping them as one individual sysex for each note). The midi notes should NOT be output in this tuning operation as the midi spec does not require them. There also needs to be some way to select the "tuning program number" for each "Tune" operation. The manufacturers ID should be 0x7f (which is used for real time messages). 0x00 is reserved for extensions and should NOT be used by itself.

csw900

PS mfrs ID comment has been updated after some research.

**Reply by midimusic** [reply]
@000masa000 
I am currently making an organ which I anticipate will eventually fully support the midi tuning standard. At present it supports only tuning program 0.

If I send you an advance copy of this program will you be ready willing and able to test the tunings and also update your hex program so that it fully supports this program's use of the midi tuning standard?

Currently I have been trying to use your program to test my organ and have found that the midi key it generates reflects the midi keyboard input only for the equal tempered scale. Others scales begin at midi key 0 no matter what keyboard key is entered.

csw900

**Comment by 000masa000** [comment]
yes i would be happy to! i have the multichannel output of mine going from 0 to allow the biggest possible scale of 128 notes to be mapped on each successive channel an octave apart! this is super exciting for me especially if it can support grandorgue sample bundles like norrfjarden church which i love

Marc Sabat

On 3. Mar 2023, at 10:30, midimusic wrote:

﻿
@000masa000
I am currently making an organ which I anticipate will eventually fully support the midi tuning standard. At present it supports only tuning program 0.

If I send you an advance copy of this program will you be ready willing and able to test the tunings and also update your hex program so that it fully supports this program's use of the midi tuning standard?

Currently I have been trying to use your program to test my organ and have found that the midi key it generates reflects the midi keyboard input only for the equal tempered scale. Others scales begin at midi key 0 no matter what keyboard key is entered.

csw900

**Comment by 000masa000** [comment]
Is there a fork of GrandOrgue I should join? 

From: midimusicSent: Friday, 3 March 2023 10:30
To: GrandOrgue/grandorgueCc: Marc Sabat; 

I am currently making an organ which I anticipate will eventually fully support the midi tuning standard. At present it supports only tuning program 0.If I send you an advance copy of this program will you be ready willing and able to test the tunings and also update your hex program so that it fully supports this program's use of the midi tuning standard?
Currently I have been trying to use your program to test my organ and have found that the midi key it generates reflects the midi keyboard input only for the equal tempered scale. Others scales begin at midi key 0 no matter what keyboard key is entered.
csw900

**Comment by midimusic** [comment]
@000masa000 
The organ I am working on is not GO - as mentioned above GO already has tuning capability which is sufficient for most organists. It would be a hige task with very little benefit to modify the code so that it could work with real time midi tuning.

My organ uses fluidsynth and was mentioned above. (You can download it from my website.) The biggest problem I currently have is testing the new capability. Your hex program - as far as I have discovered so far - is the only program potentially useable for midi tuning in accordance with the midi specs. However it falls so far short of being useful at present that I cannot use it for this purpose. You are the best person to update it and bring it into line with what is required for a useable organ tuning program. Although I understand javascript I do not want to spend my time working on your program. My favoured approach at present is to use midi files for tuning - but making these files is difficult - especially as I have just found that my midi player appears not accept 0x7f as a valid sysex mfrs. ID.

csw900

**Reply by 000masa000** [reply]
You will be pleased to hear that fluidsynth works perfectly with my app. All you need to do is give the message settuning 0 0 0 on all channels once, it will hold across any patch changes. Then if you open three copies of my app and set up three inputs for the manuals, it should relay the tunings without any further modification....

**Reply by midimusic** [reply]
@000masa000 
Yes, you will also be pleased to hear that these operations are not needed. Midi tuning messages do not distinguish one midi channel from another - they apply to all channels. Thus all manuals, Divisions and Ranks are tuned at once with one message for each midi note.

The current problem with your program is that it does not respect the note entered into it from the attached midi keyboard. For debugging it is required that if I enter middle C from the keyboard it tunes middle C (and only middle C) on the organ. It MUST NOT move on to the next note above when middle C is pressed again!!! Especially as the only way to get back to middle C is to exit the program and start again.

For tuning the whole organ - no attached keyboard should be needed. A button labelled "Tune" should output ALL the tuning messages for the selected scale  from midi note 0 to midi note 127 without stopping. There should be NO midi notes output at all.
csw900

**Comment by 000masa000** [comment]
ok i will download your program and see!

Marc Sabat


On 4. Mar 2023, at 17:11, midimusic ***@***.***> wrote:

﻿
@000masa000
The organ I am working on is not GO - as mentioned above GO already has tuning capability which is sufficient for most organists. It would be a hige task with very little benefit to modify the code so that it could work with real time midi tuning.

My organ uses fluidsynth and was mentioned above. (You can download it from my website.) The biggest problem I currently have is testing the new capability. Your hex program - as far as I have discovered so far - is the only program potentially useable for midi tuning in accordance with the midi specs. However it falls so far short of being useful at present that I cannot use it for this purpose. You are the best person to update it and bring it into line with what is required for a useable organ tuning program. Although I understand javascript I do not want to spend my time working on your program. My favoured approach at present is to use midi files for tuning - but making these files is difficult - especially as I have just found that my midi player appears not accept 0x7f as a valid sysex mfrs. ID.

csw900

**Comment by 000masa000** [comment]
Yes, true, as long as they all share the same tuning map.  ........................................Marc Sabatmobile DE +49 176 24569023mobile FI +358 41 492 ***@***.***://www.plainsound.org From: midimusicSent: Sunday, 5 March 2023 15:07To: GrandOrgue/grandorgueCc: Marc Sabat; MentionSubject: Re: [GrandOrgue/grandorgue] Integrating MIDI Tuning Standard (real-time retuning message) in GO (Discussion ***@***.***, you will also be pleased to hear that these operations are not needed. Midi tuning messages do not distinguish one midi channel from another - they apply to all channels. Thus all manuals, Divisions and Ranks are tuned at once with one message for each midi note.csw900—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by 000masa000** [comment]
However, more recent usage allows multichannel tuning maps as well, see Oddsound’s implementation using MTS-ESP, which is fast becoming a standard plug-in to translate between incoming MTS messages and receiving devices using MTS, MPE, or multichannel pitch bending to achieve tunings. .........................................Marc Sabatmobile DE +49 176 24569023mobile FI +358 41 492 ***@***.***://www.plainsound.org From: midimusicSent: Sunday, 5 March 2023 15:07To: GrandOrgue/grandorgueCc: Marc Sabat; MentionSubject: Re: [GrandOrgue/grandorgue] Integrating MIDI Tuning Standard (real-time retuning message) in GO (Discussion ***@***.***, you will also be pleased to hear that these operations are not needed. Midi tuning messages do not distinguish one midi channel from another - they apply to all channels. Thus all manuals, Divisions and Ranks are tuned at once with one message for each midi note.csw900—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by midimusic** [comment]
@000masa000 
I think I get the gist of what you are saying - however multichannel tuning maps are not needed on an organ. You must tune all manuals (i.e. midi channels), divisions and ranks to the same tuning scale or the instrument will be discordant and unmusical.
You ought to look at your posts DIRECTLY on "Github Discussions" because your replies are displayed as disjointed, confused and messy text and with unrequired HTML comments being displayed. Please edit and tidy up if you can.
csw900
